Introduce Task API

azatsarynnyy opened the pull request:
This PR closes #1698 by introducing the Task API.
One of the interesting advantages of the Task API is that it allows Theia to consume VSCode's tasks.json.

The main concepts are:
#### Task Type
A Task Type is used to define which Task Resolver, Task Provider or Task Runner should be used with a particular Task Configuration.

#### Task Resolver (frontend contribution point)
Resolves a Task Configuration before sending it for execution to the Task Server. For example, this makes it possible to hide the actual complex command line from the user.

#### Task Provider (frontend contribution point)
It allows an extension to contribute the Tasks programmatically to the system.

#### Task Runner (backend contribution point)
A Task Runner knows how to run and kill a Task of a particular type. If contribution provides a Task Runner then it will be used to run a Task of a particular type otherwise, the default Task Runner is used.
That contribution point won't be so widely used as a Task Resolver and a Task Provider. But it's necessary for the extensions like Che, whose Tasks aren't running as the processes on Theia backend.
The Task extension provides `ProcessTaskRunner` implementation that can run a Task as a process or a command inside a shell (Tasks of type `process` or `shell`). It also serves as a default runner.
***
#### Here are some demos of using the Task API
Running a Task from the VSCode sources:
1/ Just open VSCode repo as Theia workspace and copy the tasks.json from .vscode into .theia folder.
2/ Now you can run, for example, `Run tests` task.
![vscode](https://user-images.githubusercontent.com/1636395/41208374-6022f7b2-6d2b-11e8-979b-c4add8aa8c19.gif)

---
Theia npm extension (not a part of this PR) that contributes:
- [npm Task Resolver](https://github.com/theia-ide/theia/blob/artem/task-api_ext-npm/packages/npm/src/browser/npm-task-resolver.ts) to allow its own format of a Task for describing npm script;
- [npm Task Provider](https://github.com/theia-ide/theia/blob/artem/task-api_ext-npm/packages/npm/src/browser/npm-task-provider.ts) to provide the scripts detected in a package.json as the Theia Tasks.
![npm](https://user-images.githubusercontent.com/1636395/41208365-517331fa-6d2b-11e8-89d8-5debddc25867.gif)

---
Che Task extensions (not a part of this PR) that contributes:
- [Che Task Resolver](https://github.com/theia-ide/theia/blob/artem/task-api_ext-che/packages/che/src/browser/che-task-resolver.ts) to allow its own format of a Task;
- [Che Task Runner](https://github.com/theia-ide/theia/blob/artem/task-api_ext-che/packages/che/src/node/che-task-runner.ts) to execute a command in any machine (docker container) of a Che Workspace;
- [Che Task Provider](https://github.com/theia-ide/theia/blob/artem/task-api_ext-che/packages/che/src/browser/che-task-provider.ts) to provide the known Che Workspace's commands as the Theia Tasks.

![che](https://user-images.githubusercontent.com/1636395/41208361-423777a0-6d2b-11e8-8bc7-b093ba0ec489.gif)

azatsarynnyy commented:
@marcdumais-work 
Marc, could you please review that PR. I remember you also were interested in these changes )

marcdumais-work commented:
@azatsarynnyy Sounds awesome! I will have to juggle this review and a few other things, but I am interested for sure.

azatsarynnyy commented:
I've corrected the docs and changed a visibility of all properties of the TerminalWidget to protected for better extendibility.

marcdumais-work commented on packages/task/src/node/task-server.ts:
```diff
@@ -6,134 +6,60 @@
+        const taskType = taskConfiguration.type;
+        const runner = this.runnerRegistry.getRunner(taskType);
+        if (!runner) {
+            return Promise.reject(new Error(`No corresponding Runner found for the Task type ${taskType}`));
```
It might be nice to have a unit test that covers this case here, where a task runner is not found

azatsarynnyy replied:
Actually, this is a leftover from the previous version when there wasn't a default runner. I removed that code and added a test to check that a default runner is used.

marcdumais-work commented:
@azatsarynnyy  is it your intention to submit the npm task resolver and provider, show in the video above, for review next? Looks lt will be awesome for JS/TS development in Theia.

azatsarynnyy commented on packages/task/src/node/process/process-task-runner.ts:
```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (C) 2018 Red Hat, Inc. and others.
```
Yes, you're right. Those files you mentioned were actually reworked. I've changed the copyright headers to Ericsson. Thanks for noticing that.

azatsarynnyy commented:
@marcdumais-work I've addressed your last comments: updated the license headers and added a test. Also merged with the latest changes from the master branch.
> Have you had the chance to try it on Windows?

Unfortunately, I don't have a chance to try it on Windows.
Regarding my npm extension - yes, I'd like to improve it a bit and propose a PR.

azatsarynnyy commented:
Marc, thank you for taking the time to try that changes and reviewing my PR.
